File: ble-common/src/main/java/no/nordicsemi/android/ble/common/callback/ht/TemperatureTypeResponse.java
package no.nordicsemi.android.ble.common.callback.ht;

import android.bluetooth.BluetoothDevice;
import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.NonNull;

import no.nordicsemi.android.ble.exception.InvalidDataException;
import no.nordicsemi.android.ble.exception.RequestFailedException;

/**
 * Response class that could be used as a result of a synchronous request.
 * The data received are available through getters, instead of a callback.
 * <p>
 * Usage example:
 * <pre>
 * try {
 *     TemperatureTypeResponse response = readCharacteristic(characteristic)
 *           .awaitValid(TemperatureTypeResponse.class);
 *     int type = response.getType();
 *     ...
 * } catch ({@link RequestFailedException} e) {
 *     Log.w(TAG, "Request failed with status " + e.getStatus(), e);
 * } catch ({@link InvalidDataException} e) {
 *     Log.w(TAG, "Invalid data received: " + e.getResponse().getRawData());
 * }
 * </pre>
 * </p>
 */
@SuppressWarnings("unused")
public final class TemperatureTypeResponse extends TemperatureTypeDataCallback implements Parcelable {
	private int type;

	public TemperatureTypeResponse() {
		// empty
	}

	@Override
	public void onTemperatureTypeReceived(@NonNull final BluetoothDevice device, final int type) {
		this.type = type;
	}

	public int getType() {
		return type;
	}

	// Parcelable
	private TemperatureTypeResponse(final Parcel in) {
		super(in);
		type = in.readInt();
	}

	@Override
	public void writeToParcel(final Parcel dest, final int flags) {
		super.writeToParcel(dest, flags);
		dest.writeInt(type);
	}

	public static final Creator<TemperatureTypeResponse> CREATOR = new Creator<TemperatureTypeResponse>() {
		@Override
		public TemperatureTypeResponse createFromParcel(final Parcel in) {
			return new TemperatureTypeResponse(in);
		}

		@Override
		public TemperatureTypeResponse[] newArray(final int size) {
			return new TemperatureTypeResponse[size];
		}
	};
}

File: ble-common/src/main/java/no/nordicsemi/android/ble/common/callback/ht/TemperatureTypeResponse.kt
/*
 * Copyright (c) 2018, Nordic Semiconductor
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the
 * documentation and/or other materials provided with the distribution.
 *
 * 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote products derived from this
 * software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
 * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 * USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

package no.nordicsemi.android.ble.common.callback.ht

import android.bluetooth.BluetoothDevice
import android.os.Parcel
import android.os.Parcelable

import no.nordicsemi.android.ble.exception.InvalidDataException
import no.nordicsemi.android.ble.exception.RequestFailedException

/**
 * Response class that could be used as a result of a synchronous request.
 * The data received are available through getters, instead of a callback.
 *
 *
 * Usage example:
 * <pre>
 * try {
 * TemperatureTypeResponse response = readCharacteristic(characteristic)
 * .awaitValid(TemperatureTypeResponse.class);
 * int type = response.getType();
 * ...
 * } catch ([RequestFailedException] e) {
 * Log.w(TAG, "Request failed with status " + e.getStatus(), e);
 * } catch ([InvalidDataException] e) {
 * Log.w(TAG, "Invalid data received: " + e.getResponse().getRawData());
 * }
</pre> *
 *
 */
class TemperatureTypeResponse : TemperatureTypeDataCallback, Parcelable {
    var type: Int = 0
        private set

    constructor() {
        // empty
    }

    // Parcelable
    private constructor(`in`: Parcel) : super(`in`) {
        type = `in`.readInt()
    }

    override fun onTemperatureTypeReceived(device: BluetoothDevice, type: Int) {
        this.type = type
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        super.writeToParcel(dest, flags)
        dest.writeInt(type)
    }

    companion object {
        @JvmField
        val CREATOR: Parcelable.Creator<TemperatureTypeResponse> =
            object : Parcelable.Creator<TemperatureTypeResponse> {
                override fun createFromParcel(`in`: Parcel): TemperatureTypeResponse {
                    return TemperatureTypeResponse(`in`)
                }

                override fun newArray(size: Int): Array<TemperatureTypeResponse?> {
                    return arrayOfNulls(size)
                }
            }
    }
}
